File: ui/base/src/main/java/org/cru/godtools/base/ui/util/ModelUtils.java
package org.cru.godtools.base.ui.util;

import android.content.Context;
import android.graphics.Typeface;

import com.google.common.base.Strings;

import org.cru.godtools.model.Tool;
import org.cru.godtools.model.Translation;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import static com.google.common.base.MoreObjects.firstNonNull;

public final class ModelUtils {
    @NonNull
    public static CharSequence getTranslationName(@Nullable final Context context,
                                                  @Nullable final Translation translation, @Nullable final Tool tool) {
        return getTranslationName(translation != null ? LocaleTypefaceUtils.safeApplyTypefaceSpan(
                translation.getName(), getTranslationTypeface(context, translation)) : null,
                                  tool != null ? tool.getName() : null);
    }

    @NonNull
    public static CharSequence getTranslationName(@Nullable final CharSequence translationName,
                                                  @Nullable final String toolName) {
        return firstNonNull(translationName, Strings.nullToEmpty(toolName));
    }

    @NonNull
    public static CharSequence getTranslationDescription(@Nullable final Context context,
                                                         @Nullable final Translation translation,
                                                         @Nullable final Tool tool) {
        return getTranslationDescription(translation != null ? LocaleTypefaceUtils.safeApplyTypefaceSpan(
                translation.getDescription(), getTranslationTypeface(context, translation)) : null,
                                         tool != null ? tool.getDescription() : null);
    }

    @NonNull
    public static CharSequence getTranslationDescription(@Nullable final CharSequence translationDescription,
                                                         @Nullable final String toolDescription) {
        return firstNonNull(translationDescription, Strings.nullToEmpty(toolDescription));
    }

    @Nullable
    private static Typeface getTranslationTypeface(@Nullable final Context context,
                                                   @NonNull final Translation translation) {
        return LocaleTypefaceUtils.getTypeface(context, translation.getLanguageCode());
    }
}

File: ui/base/src/main/kotlin/org/cru/godtools/base/ui/util/ModelUtils.kt
@file:JvmName("ModelUtils")

package org.cru.godtools.base.ui.util

import android.content.Context
import android.content.res.Resources
import java.util.Locale
import org.ccci.gto.android.common.util.content.localizeIfPossible
import org.cru.godtools.model.Tool
import org.cru.godtools.model.Translation
import org.cru.godtools.ui.BuildConfig
import org.cru.godtools.ui.R
import timber.log.Timber

@JvmName("getTranslationName")
fun Translation?.getName(tool: Tool?, context: Context?) = this?.name ?: tool?.name ?: ""

@JvmName("getTranslationDescription")
fun Translation?.getDescription(tool: Tool?, context: Context?) = this?.description ?: tool?.description ?: ""

// region Tool Category
@JvmName("getToolCategory")
fun Tool?.getCategory(context: Context, locale: Locale? = null) = getToolCategoryName(this?.category, context, locale)
fun getToolCategoryName(category: String?, context: Context, locale: Locale? = null) =
    category?.let { c -> context.localizeIfPossible(locale).getToolCategoryStringRes(c) ?: c }.orEmpty()

private fun Context.getToolCategoryStringRes(category: String) = when (category.lowercase(Locale.ROOT)) {
    Tool.CATEGORY_GOSPEL -> getString(R.string.tool_category_gospel)
    Tool.CATEGORY_ARTICLES -> getString(R.string.tool_category_articles)
    Tool.CATEGORY_CONVERSATION_STARTERS -> getString(R.string.tool_category_conversation_starter)
    Tool.CATEGORY_GROWTH -> getString(R.string.tool_category_growth)
    Tool.CATEGORY_TRAINING -> getString(R.string.tool_category_training)
    else -> {
        val e = Resources.NotFoundException("tool_category_$category was not found")
        when {
            BuildConfig.DEBUG -> throw e
            else -> Timber.tag("ToolCategory").e(e, "Missing Tool Category string: $category")
        }
        null
    }
}
// endregion Tool Category
